include/swift/SIL/SILInstructionWorklist.h

```diff
@@ -24,92 +24,127 @@
 ///
 //===----------------------------------------------------------------------===//
 
+#include "swift/Basic/BlotSetVector.h"
 #include "swift/SIL/SILInstruction.h"
 #include "swift/SIL/SILValue.h"
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/ADT/SmallVector.h"
 
 namespace swift {
 
+class SILInstructionWorklistBase {
+  StringLiteral loggingName;
+
+protected:
+  SILInstructionWorklistBase(StringLiteral loggingName)
+      : loggingName(loggingName){};
+  void withStream(
+      std::function<void(llvm::raw_ostream &stream, StringRef loggingName)>
+          perform);
+};
+
 /// Manages a list of instructions awaiting visitation.
-class SILInstructionWorklist final {
-  llvm::SmallVector<SILInstruction *, 256> worklist;
-  llvm::DenseMap<SILInstruction *, unsigned> worklistMap;
-  StringRef loggingName;
+template <typename VectorT = std::vector<SILInstruction *>,
+          typename MapT = llvm::DenseMap<SILInstruction *, unsigned>>
+class SILInstructionWorklist : SILInstructionWorklistBase {
+  BlotSetVector<SILInstruction *, VectorT, MapT> worklist;
 
   void operator=(const SILInstructionWorklist &rhs) = delete;
   SILInstructionWorklist(const SILInstructionWorklist &worklist) = delete;
 
 public:
-  SILInstructionWorklist(const char *loggingName = "InstructionWorklist")
-      : loggingName(loggingName) {}
+  SILInstructionWorklist(StringLiteral loggingName = "InstructionWorklist")
+      : SILInstructionWorklistBase(loggingName) {}
 
   /// Returns true if the worklist is empty.
   bool isEmpty() const { return worklist.empty(); }
 
   /// Add the specified instruction to the worklist if it isn't already in it.
-  void add(SILInstruction *instruction);
+  void add(SILInstruction *instruction) {
+    if (worklist.insert(instruction).second) {
+#ifndef NDEBUG
+      withStream([&](llvm::raw_ostream &stream, StringRef loggingName) {
+        stream << loggingName << ": ADD: " << *instruction << '\n';
+      });
+#endif
+    }
+  }
 
-  /// If the given ValueBase is a SILInstruction add it to the worklist.
-  void addValue(ValueBase *value) {
-    if (auto *instruction = value->getDefiningInstruction())
+  /// If the given SILValue is a SILInstruction add it to the worklist.
+  void addValue(SILValue value) {
+    if (auto *instruction = value->getDefiningInstruction()) {
       add(instruction);
+    }
   }
 
   /// Add the given list of instructions in reverse order to the worklist. This
   /// routine assumes that the worklist is empty and the given list has no
   /// duplicates.
-  void addInitialGroup(ArrayRef<SILInstruction *> list);
+  void addInitialGroup(ArrayRef<SILInstruction *> list) {
+    assert(worklist.empty() && "worklist must be empty to add initial group");
+    worklist.reserve(list.size() + 16);
+#ifndef NDEBUG
+    withStream([&](llvm::raw_ostream &stream, StringRef loggingName) {
```

Comment:
Question: is withStream intended to ever be called if NDEBUG is defined? If so a nice little trick you can do is put the #ifndef NDEBUG into withStream. This will be a function that is pretty much guaranteed to be inlined when NDEBUG is not defined since its body will be an empty. Just a thought.

Reply:
Okay, it's now `withDebugStream` and the the `#ifndef NDEBUG` is in the body of that method in the .cpp file.